### PR TITLE
ARROW-14347: [C++] random access files for GcsFileSystem

### PR DESCRIPTION
**Pull request by coryan**

### CODE REVIEW

**Comment by coryan** [comment]
The `C++ / AMD64 Conda C++` failure seems unrelated.

File: cpp/src/arrow/filesystem/gcsfs.cc [hunk 1]
```diff
@@ -324,17 +407,46 @@ Result<std::shared_ptr<io::InputStream>> GcsFileSystem::OpenInputStream(
...
-  return impl_->OpenInputStream(p);
+  return impl_->OpenInputStream(p.bucket, p.object, gcs::Generation(),
+                                gcs::ReadFromOffset());
 }
 
 Result<std::shared_ptr<io::RandomAccessFile>> GcsFileSystem::OpenInputFile(
     const std::string& path) {
-  return Status::NotImplemented("The GCS FileSystem is not fully implemented");
+  ARROW_ASSIGN_OR_RAISE(auto p, GcsPath::FromString(path));
+  auto metadata = impl_->GetObjectMetadata(p);
```
Comment (pitrou) [inline cpp/src/arrow/filesystem/gcsfs.cc:417]: I'm curious, does this add a roundtrip to the server?

File: cpp/src/arrow/filesystem/gcsfs.cc [hunk 2]
```diff
@@ -172,6 +174,79 @@ class GcsOutputStream : public arrow::io::OutputStream {
...
+  //@}
+
+  //@{
+  // @name RandomAccessFile
+  Result<int64_t> GetSize() override { return metadata_.size(); }
+  Result<int64_t> ReadAt(int64_t position, int64_t nbytes, void* out) override {
+    std::shared_ptr<io::InputStream> stream;
+    ARROW_ASSIGN_OR_RAISE(stream, factory_(metadata_.bucket(), metadata_.name(),
+                                           gcs::Generation(metadata_.generation()),
+                                           gcs::ReadFromOffset(position)));
```
Comment (pitrou) [inline cpp/src/arrow/filesystem/gcsfs.cc:224]: Interesting. It seems a bit convoluted, but I guess the GCS C++ API is forcing us to do this.

File: cpp/src/arrow/filesystem/gcsfs_test.cc [hunk 3]
```diff
@@ -119,7 +122,23 @@ class GcsIntegrationTest : public ::testing::Test {
             .set<gc::UnifiedCredentialsOption>(gc::MakeInsecureCredentials()));
   }
 
+  std::string RandomLine(int lineno, std::size_t width) {
+    auto const fillers = std::string("abcdefghijlkmnopqrstuvwxyz0123456789");
+    std::uniform_int_distribution<std::size_t> d(0, fillers.size() - 1);
```
Comment (pitrou) [inline cpp/src/arrow/filesystem/gcsfs_test.cc:127]: You can probably reuse `random_string` from `arrow/testing/util.h` instead.

File: cpp/src/arrow/filesystem/gcsfs_test.cc [hunk 4]
```diff
@@ -521,6 +540,130 @@ TEST_F(GcsIntegrationTest, WriteObjectLarge) {
...
+
+  auto const expected = std::string(kLoremIpsum).substr(kStart);
+  EXPECT_EQ(std::string(buffer.data(), size), expected);
+}
+
+TEST_F(GcsIntegrationTest, OpenInputFileNotFound) {
+  auto fs = internal::MakeGcsFileSystemForTest(TestGcsOptions());
+
+  auto result = fs->OpenInputFile(NotFoundObjectPath());
+  EXPECT_EQ(result.status().code(), StatusCode::IOError);
```
Comment (pitrou) [inline cpp/src/arrow/filesystem/gcsfs_test.cc:650]: Can use `ASSERT_RAISES(IOError, fs->OpenInputFile(...))` here, and below as well.

File: cpp/src/arrow/filesystem/gcsfs_test.cc [hunk 5]
```diff
@@ -521,6 +540,130 @@ TEST_F(GcsIntegrationTest, WriteObjectLarge) {
...
+  auto fs = internal::MakeGcsFileSystemForTest(TestGcsOptions());
+
+  arrow::fs::FileInfo info;
+  ASSERT_OK_AND_ASSIGN(info, fs->GetFileInfo(kPreexistingBucket));
+
+  auto result = fs->OpenInputFile(NotFoundObjectPath());
+  EXPECT_EQ(result.status().code(), StatusCode::IOError);
+
+  ASSERT_OK_AND_ASSIGN(info, fs->GetFileInfo(NotFoundObjectPath()));
+  result = fs->OpenInputFile(NotFoundObjectPath());
```
Comment (pitrou) [inline cpp/src/arrow/filesystem/gcsfs_test.cc:663]: Did you mean to call `fs->OpenInputFile(info)` instead here as well?

File: cpp/src/arrow/filesystem/gcsfs_test.cc [hunk 6]
```diff
@@ -521,6 +540,130 @@ TEST_F(GcsIntegrationTest, WriteObjectLarge) {
...
+  EXPECT_EQ(result.status().code(), StatusCode::IOError);
+}
+
+TEST_F(GcsIntegrationTest, OpenInputFileInfoInvalid) {
+  auto fs = internal::MakeGcsFileSystemForTest(TestGcsOptions());
+
+  arrow::fs::FileInfo info;
+  ASSERT_OK_AND_ASSIGN(info, fs->GetFileInfo(kPreexistingBucket));
+
+  auto result = fs->OpenInputFile(NotFoundObjectPath());
```
Comment (pitrou) [inline cpp/src/arrow/filesystem/gcsfs_test.cc:659]: Did you mean to call `fs->OpenInputFile(info)` instead?

**Review by pitrou — CHANGES_REQUESTED** [review]
Some questions and a couple small comments.

File: cpp/src/arrow/filesystem/gcsfs.cc (hunk 1 above)
Comment (coryan) [reply cpp/src/arrow/filesystem/gcsfs.cc:417]: Yes, it does.  I am trying to ensure that `Read()` and `ReadAt()` and `Seek()` when going back are using the same generation of an object [*].  We could try to use undocumented (and likely to break) APIs to extract the generation without this roundtrip.  If it turns out the roundtrip (and I should add, the additional API charges) are really important, then I would rather add a documented API to the C++ client library and then use that here.


[*]: you probably know this, but objects in GCS are versioned.  You can have more than one version of the same object, and/or have the "latest" version replaced while you are reading from it.  I would think we want all operations in one `io::RandomAccessFile` to refer to the same generation.

Comment (pitrou) [reply cpp/src/arrow/filesystem/gcsfs.cc:417]: >  I would think we want all operations in one io::RandomAccessFile to refer to the same generation.

Oh, definitely. I just wonder if there's a way to extract the metadata from the first stream creation? Perhaps this is something that the GCS C++ library can provide (assuming the information exists at all at the HTTP level)?

Comment (coryan) [reply cpp/src/arrow/filesystem/gcsfs.cc:417]: And thus my (maybe too obscure) note about "undocumented (and likely to break) APIs".  The C++ client library currently returns the HTTP headers, which include thinks like `x-goog-generation` (not the full metadata). This is only intended for debugging, and I **know** of future changes that will make these headers unavailable (sometimes, when the user selects a non-HTTP transport).

I can change the C++ client to return the generation in a future-proof API.  I would rather do that in a separate PR, after I fix the C++ client library (https://github.com/googleapis/google-cloud-cpp/issues/7677).

Comment (pitrou) [reply cpp/src/arrow/filesystem/gcsfs.cc:417]: > I can change the C++ client to return the generation in a future-proof API. I would rather do that in a separate PR, after I fix the C++ client library

No problem from me.

File: cpp/src/arrow/filesystem/gcsfs_test.cc (hunk 3 above)
Comment (coryan) [reply cpp/src/arrow/filesystem/gcsfs_test.cc:127]: `random_string()` produces non-printable characters, which are not the easiest thing to troubleshoot.  `random_ascii()` could do it, but has a weird API using a `uint8_t*` output parameter.  I would rather leave this as-is.

File: cpp/src/arrow/filesystem/gcsfs_test.cc (hunk 4 above)
Comment (coryan) [reply cpp/src/arrow/filesystem/gcsfs_test.cc]: Done.

File: cpp/src/arrow/filesystem/gcsfs_test.cc (hunk 5 above)
Comment (coryan) [reply cpp/src/arrow/filesystem/gcsfs_test.cc]: Yes, thanks. Fixed.

File: cpp/src/arrow/filesystem/gcsfs_test.cc (hunk 6 above)
Comment (coryan) [reply cpp/src/arrow/filesystem/gcsfs_test.cc]: Yup, fixed too.

File: cpp/src/arrow/filesystem/gcsfs_test.cc (hunk 3 above)
Comment (pitrou) [reply cpp/src/arrow/filesystem/gcsfs_test.cc:127]: Fair enough :-)

**Review by pitrou — APPROVED** [review]

**Pull request closed without merging**